### PR TITLE
mpg, migrate fluent naming override from FluentGen

### PR DESCRIPTION
**Pull request by XiaofeiCao**
phase 1 of https://github.com/Azure/autorest.java/issues/2548

### CODE REVIEW

File: fluentnamer/src/main/java/com/azure/autorest/fluent/transformer/SchemaNameNormalization.java [hunk 1]
```diff
@@ -371,21 +371,29 @@ private String overrideName(String name) {
...
+    private boolean allUppercase(String name) {
+        return name.chars().allMatch(Character::isUpperCase);
+    }
+
+    // A partial match is when the configured keyword is only part of the whole word in the name.
+    // E.g. "lower": "Lower", and the actual name is "flower". We won't replace it to be "fLower".
+    private boolean partialMatch(String name, int index, int endIndex) {
+        return (index > 0 && isSameCase(name.charAt(index - 1), name.charAt(index)))
+            || (endIndex < name.length() && isSameCase(name.charAt(endIndex - 1), name.charAt(endIndex)));
+    }
```
Comment (XiaofeiCao) [inline fluentnamer/src/main/java/com/azure/autorest/fluent/transformer/SchemaNameNormalization.java:395]: Refactor, extract method.

File: fluentnamer/src/main/java/com/azure/autorest/fluent/transformer/SchemaNameNormalization.java [hunk 2]
```diff
@@ -371,21 +371,29 @@ private String overrideName(String name) {
...
-                    replace = false;
-                } else if (endIndex < newName.length() && isSameCase(newName.charAt(endIndex - 1), newName.charAt(endIndex))) {
-                    replace = false;
-                }
-
-                if (replace) {
+                if (!partialMatch(newName, index, endIndex)
+                    // hack, mgmt only allow `preserve-uppercase-max-length` set to 2, so there can't be a partial match
+                    // for all uppercase, e.g. IP. Otherwise, the max uppercase length is 3, which is larger than 2.
+                    // TODO(xiaofei) Remove when we done migrating m4's `preserve-uppercase-max-length`
```
Comment (weidongxu-microsoft) [inline fluentnamer/src/main/java/com/azure/autorest/fluent/transformer/SchemaNameNormalization.java:377]: I may not get what's here. One example?

Comment (XiaofeiCao) [reply fluentnamer/src/main/java/com/azure/autorest/fluent/transformer/SchemaNameNormalization.java:377]: What I thought was, with `preserve-uppercase-max-length=2`, m4 won't allow `IPRules`(which has uppercase length= 3). We don't have the m4 restriction yet. Thus we can safely assume, when a name has "IP" in it, it's the max uppercase length and the whole word, and we will replace it.

Though through double check, `preserve-uppercase-max-length` seems not working as I assumed.. May need revisit this hack myself..

Comment (weidongxu-microsoft) [reply fluentnamer/src/main/java/com/azure/autorest/fluent/transformer/SchemaNameNormalization.java:377]: My guess is m4 first breaks `IPRules` to "ip" + "rules", then check max-uppercase. So it be 2, not 3. Not sure.

Comment (XiaofeiCao) [reply fluentnamer/src/main/java/com/azure/autorest/fluent/transformer/SchemaNameNormalization.java:377]: Got it.. Then it seems m4 has different naming override strategy with us..

Comment (weidongxu-microsoft) [reply fluentnamer/src/main/java/com/azure/autorest/fluent/transformer/SchemaNameNormalization.java:377]: I think so. m4 logic is more complete. My logic here is not intended for uppercase or lowercase and probably not work for short words.

I guess you can ignore ID IP for now.

Comment (XiaofeiCao) [reply fluentnamer/src/main/java/com/azure/autorest/fluent/transformer/SchemaNameNormalization.java:377]: Sure, will revisit it later.

**Review by weidongxu-microsoft — APPROVED** [review]

**Pull request merged**